<a href="https://colab.research.google.com/github/Elensk8/Anal-tica_Neg/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargar librerías

In [ ]:
import numpy as np
import pandas as pd

#Para implementación del modelo
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

from sklearn.preprocessing import LabelEncoder

1. Cargar base de datos

In [ ]:
nxl = '/content/2. TechSalesReport_Int_L.xlsx'
XDB = pd.read_excel(nxl,sheet_name=0)


2. Llamar solo las variables que necesito

In [ ]:
df2 = XDB[['Business', 'Age', 'Female', 'Years','Certificates', 'Feedback','Salary','NPS','College']]

3. Hacer que las variables Business y College sean numéricas

In [ ]:
df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})

<ipython-input-23-2fda56836961>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Business'] = df2['Business'].map({'Hardware': 0, 'Software': 1})
<ipython-input-23-2fda56836961>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['College'] = df2['College'].map({'Yes': 1, 'No': 0})


4. Cargar las entradas y salida

In [ ]:
#Seleccionamos las variables
XD=df2[['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']]
yd=df2[['College']]

5. Se procede con la implementación del modelo

*Max Depth es el número de variables disponibles

*Gini es el criterio que indica el grado de error en el modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)

mar.fit(XD,yd)

#Se procede con el despliegue del árbol de decisión

from sklearn.tree import export_graphviz #Exportar
from pydotplus import graph_from_dot_data #Crear el gráfico

vs=['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']
dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('1.TechSalesReport.png')



True

6. Matriz de confusión

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]


# Total de datos
total_datos = np.sum(cm)
print("Total de datos:", total_datos)

# Datos clasificados como 'College' (positivos)
total_college = VP + FN
print("Cantidad que fueron a College:", total_college)

# Datos clasificados como 'No College' (negativos)
total_no_college = VN + FP
print("Cantidad que no fueron a College:", total_no_college)

Exactitud=(VP+VN)/(VP+VN+FP+FN) #Funcionamiento general del modelo
print('La exactitud:\n',Exactitud)
Sensibilidad=VP/(VP+FN) #Cómo se comporto el modelo con los Aprobados
print('La sensibilidad:\n',Sensibilidad)
Especificidad=VN/(VN+FP) #Cómo se comporto el modelo con los Negados
print('La especificidad:\n',Especificidad)
Precisión=VP/(VP+FP) #Cuál es el acierto clasificando los positivos
print('La precisión:\n',Precisión)
TasadeError=(FP+FN)/(VP+VN+FP+FN) #Equivocación general del modelo
print('La tasa de Error:\n',TasadeError)
Predneg=VN/(VN+FN) #Como funciona el modelo negando créditos
print('La tasa de predicción nengativa es:\n',Predneg)

[[  114  4356]
 [   69 17451]]
Total de datos: 21990
Cantidad que fueron a College: 17520
Cantidad que no fueron a College: 4470
La exactitud:
 0.7987721691678036
La sensibilidad:
 0.9960616438356165
La especificidad:
 0.025503355704697986
La precisión:
 0.8002476269087908
La tasa de Error:
 0.20122783083219645
La tasa de predicción nengativa es:
 0.6229508196721312


7. Pronóstico

In [ ]:
# Leer el archivo Excel
XDB2 = pd.read_excel(nxl, sheet_name=1)

# Convertir variables categóricas a numéricas
XDB2['Business'] = XDB2['Business'].map({'Hardware': 0, 'Software': 1})
XDB2['College'] = XDB2['College'].map({'Yes': 1, 'No': 0})

# Definir las variables predictoras
XD2 = XDB2[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']]

# Hacer predicciones
ydp2 = mar.predict(XD2)
print("Las predicciones son: \n", ydp2)

# Crear el informe con los resultados
dfp = XDB2[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']].copy()
dfp['College_Predicted'] = ydp2  # Agregar la predicción
dfp.insert(0, 'Index', XDB2.index)  # Agregar el índice original

# Guardar en Excel
dfp.to_excel('/content/2. TechSalesReport_Int_L.xlsx', index=False)

# Mostrar conteo de predicciones
print("Los negados pronosticados son:\n", len(np.where(ydp2 == 0)[0]))
print("Los aprobados pronosticados son:\n", len(np.where(ydp2 == 1)[0]))

Las predicciones son: 
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
Los negados pronosticados son:
 0
Los aprobados pronosticados son:
 16


Análisis:
Por medio de la base de datos de Tech Sales Report, se busca de definir a través de un modelo de árbol de decisión,si los empleados poseen o no un título universitario (College) (nuestra variable de salida), en base a ciertas variables de salida, y cuáles son las más relevantes de ellas para pronosticar si estudio College o no el trabajador. Para esto, se utilizaron variables como: Business (si se dedicaban al hardware o al software), Age (edad del representante de ventas) y Female (para identificar si es hombre o mujer) que son dos variables no congitivas las cuáles querían implementar y analizar que relacion tenían, Years, Certificates (número de certificados que tienen), Feedback (puntajes), Salary que es visto como indicador de éxito laboral y NPS.
Se estudiaron en total 21990 datos, de los cuales 17520 si estudiaron en College y 4470 no. Adicionalmente, para formular el árbol se tomáron 4 niveles de los cuales ningúno puede ser considerado nodo puro en base a la estructura del modelo.

Por otro lado, se hico un pronóstico de aquellos Sales_Rep que no tenían información sobre el estudio de College de algunos de los trabajadores, y en base a los resultados, podemos decir que los 16 trabajadores que se analizaron, en base a los resultados del modelo es posible concluir que todos fueron a College, la cuestión es que hay un alto nivel de predominancia en la cantidad de datos de trabajadores que van a College (17520 de 21990), lo que hace que la especificidad del modelo sea bastante baja (0.025).

Por último, en la evaluación de las métricas del desempeño del modelo nos encontramos con lo sigiente: la exactitud del modelo es cercana al 80%, la especificidad es del 025%, muy bajita como ya se menciono anteriormente, la sensibilidad del 99%, la precisión del 80%, la tasa de error del 20% y la tasa de predicción negativa de más del 62%. Los verdaderos negativos son 114, mientras que los verdaderos positivos son 17451, los falsos positivos 4356, y los falsos negativos fueron apenas 69. Con esto podemos concluir que el modelo funciono bastante bien, sin embargo, si se quisiera mejorar un poco, especialmente para aumentar el porcentaje de especificidad, se debería de sacar una población aleatoria menor de los 17520 datos de trabajadores que fueron a College.

(Para revisar el resultado del pronóstico, abrir el excel anexo, el cual contiene a la columna de College con los trabajadores que fueron (1)).